cmd/aro/rp.go

```diff
@@ -63,42 +66,114 @@ func rp(ctx context.Context, log *logrus.Entry) error {
+		return err
+	}
+
+	cipher, err := encryption.NewXChaCha20Poly1305(ctx, dbKey)
+	if err != nil {
+		return err
+	}
+
+	dbc, err := database.NewDatabaseClient(ctx, log.WithField("component", "database"), _env, m, cipher)
+	if err != nil {
+		return err
+	}
 
-	cipher, err := encryption.NewXChaCha20Poly1305(ctx, _env, env.EncryptionSecretName)
+	dbasyncoperations, err := database.NewAsyncOperations(_env, dbc)
 	if err != nil {
 		return err
 	}
 
-	db, err := database.NewDatabase(ctx, log.WithField("component", "database"), _env, m, cipher, uuid)
+	dbbilling, err := database.NewBilling(ctx, _env, dbc)
 	if err != nil {
 		return err
 	}
 
-	go db.EmitMetrics(ctx)
+	dbopenshiftclusters, err := database.NewOpenShiftClusters(ctx, _env, dbc, uuid)
+	if err != nil {
+		return err
+	}
 
-	feCipher, err := encryption.NewXChaCha20Poly1305(ctx, _env, env.FrontendEncryptionSecretName)
+	go database.EmitQueueLengthMetrics(ctx, log.WithField("component", "database"), dbopenshiftclusters, m)
+
+	dbsubscriptions, err := database.NewSubscriptions(ctx, _env, dbc, uuid)
 	if err != nil {
 		return err
 	}
 
-	f, err := frontend.NewFrontend(ctx, log.WithField("component", "frontend"), _env, db, api.APIs, m, feCipher, adminactions.New)
+	feKey, err := kv.GetSecret(ctx, env.FrontendEncryptionSecretName)
+	if err != nil {
+		return err
+	}
+
+	feCipher, err := encryption.NewXChaCha20Poly1305(ctx, feKey)
+	if err != nil {
+		return err
+	}
+
+	dialer, err := proxy.NewDialer(_env)
+	if err != nil {
+		return err
+	}
+
+	var armClientAuthorizer, adminClientAuthorizer clientauthorizer.ClientAuthorizer
+	var l net.Listener
+	if _env.Type() == env.Dev {
+		armClientAuthorizer = clientauthorizer.NewAll()
+		adminClientAuthorizer = clientauthorizer.NewAll()
+
+		// in dev mode there is no authentication, so for safety we only listen on
+		// localhost
+		l, err = net.Listen("tcp", "localhost:8443")
+		if err != nil {
+			return err
+		}
+
+	} else {
+		armClientAuthorizer = clientauthorizer.NewARM(log)
+		adminClientAuthorizer, err = clientauthorizer.NewAdmin(
+			log,
+			"/etc/aro-rp/admin-ca-bundle.pem",
+			os.Getenv("ADMIN_API_CLIENT_CERT_COMMON_NAME"),
+		)
+		if err != nil {
+			return err
+		}
+
+		l, err = net.Listen("tcp", ":8443")
+		if err != nil {
+			return err
+		}
+	}
+
+	key, certs, err := kv.GetCertificateSecret(ctx, env.RPServerSecretName)
+	if err != nil {
+		return err
+	}
+
+	l = frontend.TLSListener(l, key, certs)
+
+	f, err := frontend.NewFrontend(ctx, log.WithField("component", "frontend"), _env, dialer, dbasyncoperations, dbopenshiftclusters, dbsubscriptions, l, api.APIs, m, feCipher, adminactions.New, armClientAuthorizer, adminClientAuthorizer)
 	if err != nil {
 		return err
 	}
 
-	b, err := backend.NewBackend(ctx, log.WithField("component", "backend"), _env, db, m)
+	b, err := backend.NewBackend(ctx, log.WithField("component", "backend"), _env, dialer, dbasyncoperations, dbbilling, dbopenshiftclusters, dbsubscriptions, cipher, m)
```

Comment:
we have seen problems with passing heaps of arguments into constructs and moved away from them.. this is going the other way